tests/hypermodel/test_preprocessor.py

```diff
@@ -124,6 +125,50 @@ def map_func(x):
     assert isinstance(new_dataset, tf.data.Dataset)
 
 
+def test_feature_engineering():
+    data = structured_data()
+    dataset = tf.data.Dataset.from_tensor_slices(data)
+    feature = preprocessor.FeatureEngineering([
+        'categorical', 'categorical', 'numerical', 'numerical', 'categorical',
+        'categorical', 'categorical', 'categorical'])
```

Comment:
This API is error prone and does not scale well to hundreds of features.

A better API: types are inferred most of the time (so the user doesn't specify anything). In cases where the user wants to treat e.g. an int column as categorical, the user passes `{column_name: 'categorical'}`

Reply:
Yes, we can refactor a little bit to infer the column types. however, does tf.data.Dataset has a header of column names?

Reply:
It depends on what you intend to be the input to the pipeline.

- A raw CSV file? In that case we could use the header.
- A tf.data Dataset created with `make_csv_dataset`? In that case it yields a dictionary and therefore contains column names. [Ref](https://www.tensorflow.org/beta/tutorials/load_data/csv)
- A Dataset or array structured as a list of values? In that case it won't contain column name information, and it would have to be passed via a different mechanism (e.g. a `header` argument in the constructor, which would be a list of strings)

I think it would be reasonable to require a dict input (dict of arrays or Dataset that yield dicts), in which case the column name information is part of the data. Would this choice create any issues?

Reply:
By dict input, do you mean each column name is a key, where the values are arrays (for the dict of array case) and the values are single values while iterating the Dataset?

I think this solution is good. It means we will support different types of input for different type of input nodes. I don't think we can support dict input for image input node.

I am intended to convert whatever input type from the user (csv, dataset, and etc) to a tf.Dataset with a header. Do you think it should be good?

Reply:
Yes, I think being able to accept the raw data from the user, in particular CSV files, is a good idea.